### PR TITLE
[SE-0211] Add Unicode properties to Unicode.Scalar

### PR DESCRIPTION
**Pull request by allevato**
This PR is a work-in-progress for the `Unicode.Scalar` additions being proposed in [this thread](https://forums.swift.org/t/adding-unicode-properties-to-unicodescalar-character/9310). I'm opening the PR now to link to it in the proposal text.

I still need to test these changes on Linux and add some unit tests (and likely some benchmarks) for it. Please feel free to comment on these changes—I'm very interested in that feedback—but let's hold off on running the CI bots to preserve resources until I've completed those tasks.

### CODE REVIEW

**Comment by milseman** [comment]
The UCD also defines a `isNewline` property. Do you know if ICU exposes this?

http://www.unicode.org/Public/UCD/latest/ucd/auxiliary/WordBreakProperty.txt

**Comment by allevato** [comment]
@milseman: 
> The UCD also defines a `isNewline` property. Do you know if ICU exposes this?

This enumeration is exposed via the `U_CHAR_WORD_BREAK` property. I don't believe this is a "isNewline" property as such (i.e., not a binary property), but just one case in the enumeration.

What's a bit tricky is that the `Newline` word break type *excludes* CR and LF, because those are represented as distinct word break types (whereas, conceptually, the word break algorithm in [UAX #29](https://www.unicode.org/reports/tr29/#WB3a) treats them equivalently).

So we have a couple options:

1. We could expose a computed `isNewline` property, but we'd have to decide whether we also want to include CR and LF or not; either way, we'd be diverging slightly from the standard.
1. We can just expose a `Unicode.WordBreak` enum through a `wordBreak` property and let users decide how to handle it. That's closest to standard compliance.

WDYT? Given how closely we've adhered to the standard elsewhere, I think it's hard to rationalize doing anything other than 2.

**Comment by milseman** [comment]
Oh yuck. This enum seems more like an ICU internal detail than a Unicode definition, so I'd say omit it.

**Comment by allevato** [comment]
Oops, I have to correct myself: `UCHAR_WORD_BREAK` returns `UWordBreakValues`—they had to use a different name because they had already used `UWordBreak` for their break iterator constants.

It looks like [`UWordBreakValues`](http://icu-project.org/apiref/icu4c/uchar_8h.html#af70ee907368e663f8dd4b90c7196e15c) corresponds _almost_ to the `WordBreakProperty.txt` Unicode property definitions defined in [Table 3 of UAX #29](https://www.unicode.org/reports/tr29/#Table_Word_Break_Property_Values). The ICU enum is a subset—for example, `MidLetter` doesn't have a corresponding ICU constant.

I'd have to do some more digging to see why they're not an exact match.

**Comment by milseman** [comment]
@swift-ci please test compiler performance

File: stdlib/public/core/UnicodeScalarProperties.swift
Comment (milseman) [inline stdlib/public/core/UnicodeScalarProperties.swift]: Use `String._fromUTF8CodeUnitSequence/_fromWellFormedUTF8CodeUnitSequence/_fromASCI/`, which will attempt to form a small string if possible. Try to avoid creating a storage class.

You could make a small string and try forming into that, falling back to an Array+`_fromASCII` otherwise. See e.g. `_SmallString._withMutableExcessCapacityBytes`. Micro-optimizing further is almost certainly not worth it here.

File: stdlib/public/core/UnicodeScalarProperties.swift
Comment (milseman) [inline stdlib/public/core/UnicodeScalarProperties.swift]: And it's not a `Character` because it can change a single scalar into multiple graphemes.

File: stdlib/public/core/UnicodeScalarProperties.swift
Comment (milseman) [inline stdlib/public/core/UnicodeScalarProperties.swift]: We definitely want to avoid creating storage for these.

File: stdlib/public/core/UnicodeScalarProperties.swift
Comment (milseman) [inline stdlib/public/core/UnicodeScalarProperties.swift]: Encoding a scalar into UTF-16 is pretty trivial and almost certainly not worth storing an extra word or two to avoid re-computing it. It's also not worth doing every time a user writes `myScalar.properties.isWhitespace`. I suggest computing this lazily.

**Review by milseman — COMMENTED** [review]
LGTM for a prototype, some small adjustments

File: stdlib/public/core/UnicodeScalarProperties.swift
Comment (milseman) [reply stdlib/public/core/UnicodeScalarProperties.swift]: The good news is that this code doesn't compile anymore, so it's a great time to fix it!

```swift
/Users/buildnode/jenkins/workspace-private/swift-PR-compiler-performance-macOS/master/new/swift/stdlib/public/core/UnicodeScalarProperties.swift:1091:12: error: argument labels '(_storage:)' do not match any available overloads
16:35:05     return String(_storage: storage)
```

Comment (milseman) [reply stdlib/public/core/UnicodeScalarProperties.swift]: I have a fix for case conversion (that uses a stack buffer). I'll try to do _scalarName too and share that

Comment (allevato) [reply stdlib/public/core/UnicodeScalarProperties.swift]: I've fixed `_scalarName` and pushed it—thanks for the pointers to the new APIs.

I hadn't gotten to case conversion yet since those need to be written into UTF-16 buffers and I didn't see a small analogue of `_SmallUTF8String`, so I figured there would be some extra work that I haven't been able to dive into yet. If you already have a fix, I'll just wait for that!

File: stdlib/public/core/UnicodeScalarProperties.swift
Comment (allevato) [reply stdlib/public/core/UnicodeScalarProperties.swift]: Updated.

File: stdlib/public/core/UnicodeScalarProperties.swift
Comment (allevato) [reply stdlib/public/core/UnicodeScalarProperties.swift]: Sounds good. Since you said above that you're working on a fix for the case mappings, I'll hold off on fixing this until that's ready.

Comment (milseman) [reply stdlib/public/core/UnicodeScalarProperties.swift]: Do you think you could cherry-pick this: https://github.com/milseman/swift/commit/0fdd3d79dc33bc55798de392af2aa096c59563c9 ?

I tried to open a PR against this PR, but I couldn't find your repo in the list...

Comment (milseman) [reply stdlib/public/core/UnicodeScalarProperties.swift]: Do you have any tests? I want to make sure I didn't break anything

Comment (allevato) [reply stdlib/public/core/UnicodeScalarProperties.swift]: Thanks! These new case mappings look like they did the trick (I've pushed an update with your changes, and also removed the UTF-16 caching).

I don't have any proper unit tests, I've just been spot-testing with the REPL. I'll get those together next.

**Comment by milseman** [comment]
@swift-ci please test compiler performance

**Comment by milseman** [comment]
(I measured about a 1% size increase to optimized stdlib binary size)

**Comment by milseman** [comment]
@swift-ci please smoke test compiler performance

File: stdlib/public/core/UnicodeScalarProperties.swift
Comment (allevato) [reply stdlib/public/core/UnicodeScalarProperties.swift]: 👍  Updated with your version.

**Comment by allevato** [comment]
I'll work on some tests next and get those pushed.

**Comment by milseman** [comment]
@swift-ci please smoke test compiler performance

**Comment by allevato** [comment]
I've updated some tests for the properties with more complex logic (like case mappings and age); I plan to cover the remainder shortly.

Any idea what the compiler performance failures are about, since there are no logs here?

I still need to verify that this builds on Linux, so I wonder if that has any relation.

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
https://ci.swift.org/view/Pull%20Request/job/swift-PR-compiler-performance-smoke-test-macOS/214/consoleFull#-10877886173122a513-f36a-4c87-8ed7-cbc36a1ec144

```
/Users/buildnode/jenkins/workspace-private/swift-PR-compiler-performance-smoke-test-macOS/master/new/swift/stdlib/public/core/UnicodeScalarProperties.swift:1089:17: error: value of type '_SmallUTF8String' has no member '_withAllUnsafeMutableBytes'
15:50:41     let count = smol._withAllUnsafeMutableBytes { bufPtr -> Int in
15:50:41                 ^~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~
```

That is... weird if it's on 64-bit, but makes sense on 32-bit. This would need to be `#if` guarded for 32-bit.

For case-conversion, I think it's better to just form a String and call a method on that. SmallString will support UTF-8 in the future, so that would include all scalars in the small form anyways.

For the name of a scalar, I think it's better to use a fixed size array and call String._fromASCII on that. I can try a patch to this effect later.

**Comment by allevato** [comment]
@milseman I pushed some updates now that SE-0211 has been accepted. `numericValue` returns `nil` instead of `.nan` for non-numerics, I removed the small-string optimizations from `_scalarName` that were causing 32-bit issues up above, I moved the case mappings back into `Unicode.Scalar.Properties` with names derived from the full Unicode property name (i.e., `var Unicode.Scalar.Properties.lowercaseMapping` instead of `func Unicode.Scalar.lowercased()`).

PTAL and let me know if there's anything I've missed. Thanks!

**Review by milseman — APPROVED** [review]
LGTM

**Comment by milseman** [comment]
@swift-ci please test

**Comment by allevato** [comment]
Did the tests actually run? I couldn't dig up any results on Jenkins a few hours after you kicked it off.

**Comment by milseman** [comment]
@swift-ci please test

**Comment by allevato** [comment]
Urgh, that test failure is gross.

`UCHAR_EMOJI` and the other three properties we expose were added in [ICU 57](http://icu-project.org/apiref/icu4c/uchar_8h.html#ae40d616419e74ecc7c80a9febab03199afb489371bccc9d5023d370969f1a9dec), but as far as I can tell, Ubuntu 16.04 comes with ICU 55. So the return value for those properties can't be relied on there.

The cheap way to fix it is to remove those tests, but that just hides the underlying issue. I can think of a couple options, both kind of awful:

1) Provide a way for users to either explicitly check the ICU version that stdlib was built against, but then users have to also be able to discover if a property is supported by that version. That's a lot of boilerplate code.

2) Add some conditional compilation directives to `UnicodeScalarProperties.swift` that only defines properties that actually exist based on which version of ICU is being included. That effectively introduces dialects of the standard library based on what OS you're compiling on, which would be horrible.

What are your thoughts?

**Comment by milseman** [comment]
It's really unfortunate that these APIs aren't available on the oldest Ubuntu TLS. Let me check up on what we can do here. For now, I would conditionally test based either on Darwin or a recent Linux

**Comment by allevato** [comment]
Sounds good—I've updated the PR to only define the four emoji properties on Darwin OSes.

**Comment by milseman** [comment]
@swift-ci please test

**Comment by allevato** [comment]
Tests look good! Anything left to do before we merge?

**Comment by gottesmm** [comment]
@allevato Great job!

File: stdlib/public/core/UnicodeScalarProperties.swift
Comment (AliSoftware) [inline stdlib/public/core/UnicodeScalarProperties.swift]: I thought the discussion agreed on making that enum RawRepresentable (which would make it useful for building regular expressions for example)?

https://forums.swift.org/t/se-0211-add-unicode-properties-to-unicode-scalar/12121/6

Comment (allevato) [reply stdlib/public/core/UnicodeScalarProperties.swift]: Since the proposal was accepted by the core team with the only revision being to the `numericValue` property, I didn't make any changes from the original proposal other than the one that was requested (and to remove a couple ICU-only properties that accidentally got into the list).

@milseman What do you think? Did the core team have any feelings about that specifically, or was it just an idea that just got lost in the discussion thread?

Comment (AliSoftware) [reply stdlib/public/core/UnicodeScalarProperties.swift]: Btw I don't think it would be _that_ useful to be able to do in the direction of `init(rawValue: String)` using `"Lu"` for example, but would _definitely_ be useful to go in the other direction (`. uppercaseLetter` => `"Lu"`). So only implementing `var description: String` for that purpose (instead of making it `RawRepresentable where RawValue: StringProtocol`) could also be enough

Comment (milseman) [reply stdlib/public/core/UnicodeScalarProperties.swift]: There's many ways of describing it. There's the short name, the long name, how it might appear in various regex conventions, etc. I don't have any strong opinions here, other than `description` and `debugDescription` probably should choose user-readable names over  a particular regex convention.

Long term, Swift built in regexes and/or regex packages of various flavors could provide their `Custom[PCRE|POSIX|...]RegexConvertible` protocols. The would probably produces types that themselves could be `CustomStringConvertible`, producing e.g. `"\\p{Lu}"` or `"<:Lu>"`, etc.

**Pull request merged**
